core/project_settings.cpp

```diff
@@ -390,6 +407,8 @@ Error ProjectSettings::setup(const String &p_path, const String &p_main_pack, bo
 	if (resource_path.length() && resource_path[resource_path.length() - 1] == '/')
 		resource_path = resource_path.substr(0, resource_path.length() - 1); // chop end
 
+	_convert_to_last_version();
```

Comment:
Well, no, I'm quite sure I will end up with as many hacks with this solution too, I have to check for a type in any case so it's not a problem. 
But actually, I quite disagree on the fact that this is a hack, it just converts the projects settings to the last version on startup. I think it's quite a nicer and and a lot easier to maintain having a place dedicated to project settings version update. But maybe I can have @akien-mga 's point of view on this one

Reply:
The logic for not hacking this here is the general rule of hacks I always applied to Godot.

Hacks should be always local to the place they are needed. This one is in a completely unrelated place.

Reply:
Yeah, but here I don't consider this as a hack. It's just that we consider that we have a new version of the project settings file, and add a converter from one version from another. What changed between the version is completely self documented into this function. And we're done.

My point is that the current project settings are probably not going to stay like that in the future. Adding a proper version handling here makes total sense to me.

Reply:
I admit I didn't think of it that way, as I mentioned, the version number was added there more for the format itself rather than changes in the configuration itself.  The way you are using it makes sense, though, so I guess it's ok to use it like this.